lib/pbench/cli/server/audit.py

```diff
@@ -0,0 +1,187 @@
+                f"    ID {audit.id}, ROOT {audit.root_id}: OBJ "
+                f"{audit.object_id}, UID {audit.user_id}\n"
+            )
+        if audit.attributes:
+            yield f"      {audit.attributes}\n"
+    yield ""
+    yield f"Reported {count:,d} audit events:\n"
+    for s, c in status.items():
+        yield f"    {s.name:>10s} {c:>10,d}\n"
+    if open:
+        yield f"{len(open):,d} unterminated events:\n"
+        for audit in open.values():
+            yield (
+                f"  [{audit.timestamp:%Y-%m-%d %H:%M:%S}] {audit.id:10d} "
+                f"{audit.name} {audit.object_name} ({audit.user_name}) {audit.attributes}\n"
+            )
+
+
+@click.command(name="pbench-audit")
+@pass_cli_context
+@click.option(
+    "--ids",
+    default=False,
+    is_flag=True,
+    help="Show user and object IDs as well as names",
+)
+@click.option("--name", type=str, help="Select by audit event name")
+@click.option(
+    "--operation",
+    type=click.Choice([o.name for o in OperationCode], case_sensitive=False),
+    help="Select by audit operation name",
+)
+@click.option(
+    "--object-type",
+    type=click.Choice([t.name for t in AuditType], case_sensitive=False),
+    help="Select by object type",
+)
+@click.option("--object-id", type=str, help="Select by object ID")
+@click.option("--object-name", type=str, help="Select by object name")
+@click.option("--page", default=False, is_flag=True, help="Paginate the output")
+@click.option("--user-id", type=str, help="Select by user ID")
+@click.option("--user-name", type=str, help="Select by username")
+@click.option(
+    "--status",
+    type=click.Choice([s.name for s in AuditStatus], case_sensitive=False),
+    help="Select by operation status",
+)
+@click.option(
+    "--summary", default=False, is_flag=True, help="Show one-line summary of operations"
+)
+@click.option(
+    "--since",
+    type=click.DateTime(),
+    help="Select entries on or after specified date/time",
+)
+@click.option(
+    "--until",
+    type=click.DateTime(),
+    help="Select entries on or before specified date/time",
+)
+@click.option(
+    "--verify", "-v", default=False, is_flag=True, help="Display intermediate messages"
+)
+@common_options
+def audit(context: object, **kwargs):
+    """Query and format the audit DB table
+
+    The Audit table records a sequence of event representing all changes made
+    to the data controlled by the Pbench Server. This tool supports queries to
+    display audit events based on various search criteria including timestamp,
+    user, object identification, and others.
+    \f
+
+    Args:
+        context: click context
+        kwargs: click options
+    """
+    global verifier
+    verifier = Verify(kwargs.get("verify"))
+
+    try:
+        config_setup(context)
+        if kwargs.get("page"):
+            click.echo_via_pager(auditor(kwargs))
+        else:
+            for line in auditor(kwargs):
+                click.echo(line, nl=False)
```

Comment:
Rather than calling `echo()` for each line while suppressing the newline, consider calling it once with the whole output:
```python
            click.echo(''.join(auditor(kwargs))
```

Reply:
Maybe that'd work, but it'd be enormous. Remember that `auditor()` is a generator and we can dump hundreds of thousands of lines...

Reply:
Yeah, I thought about that a little bit before suggesting, and I decided that even terminal operations could withstand large transfers (and that any really significant transfers would probably be paginated or piped or otherwise not care).

Reply:
Instant is good.